File: apps/desktop/src/modules/db/handoff.ts
import { invoke } from "@tauri-apps/api/core";
import type { ConnectionConfig } from "./types";

/**
 * A connection another program handed to MixDB — MixEngine's `mix database open` — as it is
 * taken by the tab opened for it.
 *
 * The backend read it off a `mixdb://connect?…` URL and the password off one environment
 * variable, and keeps it until this call; see
 * `docs/superpowers/specs/2026-09-03-mixengine-connection-handoff-design.md`. From here it is a
 * config in the form like any typed by hand: `connect_db`, the workspace, Save if the user wants.
 */
export interface Handoff {
  config: ConnectionConfig;
  /** The tab's name and the name pre-filled for saving — the launcher's own id for the server. */
  label: string;
  /** The address to save instead of a copy of the password — see `SavedConnection.keyringRef`.
   *  `null`/absent for a `mixdb://` link that arrived without proof MixEngine started this
   *  process, even when the URL itself named one. */
  keyring_ref?: string | null;
}

/** Takes the handoff under `id`. Rejects with `error.handoffExpired` when it was already taken,
 *  or when the id is from a session an earlier run wrote — in which case the tab is simply empty. */
export function takeHandoff(id: string): Promise<Handoff> {
  return invoke("handoff_take", { id });
}

File: apps/desktop/src-tauri/src/modules/db/commands/handoff.rs
//! The one command the handoff needs: the tab opened for it takes what was handed over, once.

use tauri::State;

use crate::error::AppError;
use crate::modules::db::handoff::{Handoff, HandoffState};

/// The handoff under `id`, removed as it is handed out. A second call, or an id from a session
/// written by an earlier run, is `error.handoffExpired` — which the tab answers with an empty
/// form and no banner, since nothing the user did has failed.
#[tauri::command]
pub fn handoff_take(state: State<'_, HandoffState>, id: String) -> Result<Handoff, AppError> {
    state.take(&id).ok_or_else(|| err!("error.handoffExpired"))
}
